File: iPad/VerBank-CSTSv3-IOSClientApi/CSTS3Proxy/Comm/CSTSTradeWaitObject.m
//
//  CSTSTradeWaitObject.m
//  JEDIv7-CSTSv3-IOSClientApi
//
//  Created by felix on 7/17/13.
//  Copyright (c) 2013 allone. All rights reserved.
//

#import "CSTSTradeWaitObject.h"
#import "JEDISystem.h"
#import "IPOPErrCodeTable.h"


@implementation CSTSTradeWaitObject

//--------------------------------------------------------------------------
- (id) init
{
    self = [self init];
    
    if(self != nil)
    {
        mIp = nil;
        mOp = nil;
        mHashCode = nil;
        mWaitLock = [[NSLock alloc] init];
    }
    
    return self;
}

//--------------------------------------------------------------------------
- (id) initWithIp:(IPFather *) ip
{
    self = [super init];
    
    if(self != nil)
    {
        mIp = ip;
        mOp = nil;
        mHashCode = mIp ? [mIp getOperateId] : nil;
        mWaitLock = [[NSLock alloc] init];
    }
    
    return self;
}

//--------------------------------------------------------------------------
- (void) dealloc
{
    mHashCode = nil;
    mWaitLock = nil;
    mIp = nil;
    mOp = nil;
}

//--------------------------------------------------------------------------
- (OPFather *) waitTradeForTime:(long) waitTime
{
    if(mOp != nil){
        return mOp;
    }
    
    @try
    {
        [mWaitLock lock];
        
        NSDate *nsDate = [NSDate dateWithTimeIntervalSinceNow:(waitTime/1000.0)];
        if(![mWaitLock lockBeforeDate:nsDate]){
            // time out for wait
        }
        
        [mWaitLock unlock];
    }
    @catch (NSException *exception) {
        [JEDISystem log:@"CSTSTradeWaitObject.waitTradeForTime, Error: " withObject:exception];
    }
    
    if(mOp == nil){
        OPFather * op = [[OPFather alloc] initWithIp:mIp];
        [op setSuccess:NO];
        [op setErrCode:ERR_Timeout]; //IPOPErrCodeTable.ERR_Timeout
        [op setErrMessage:@"CSTSTrade timeout!"];
        return op;
    }else{
        return mOp;
    }
}

//--------------------------------------------------------------------------
- (OPFather *) waitTrade
{
    return [self waitTradeForTime:JEDI_CSTS_TRADE_WAIT_GAP];
}

//--------------------------------------------------------------------------
- (void) tradeReturn:(OPFather *)op
{
    mOp = op;
    
    if(mWaitLock != nil){
        [mWaitLock unlock];
    }
}

//--------------------------------------------------------------------------
- (void) setError
{
    if(mOp != nil)
    {
        if(mWaitLock) [mWaitLock unlock];
    }
    else
    {
        mOp = [[OPFather alloc] initWithIp:mIp];
        
        [mOp setSuccess:NO];
        [mOp setErrCode:ERR_NetErr];
        [mOp setErrMessage:@"CSTSTrade network erro!"];
        [self tradeReturn:mOp];
    }
}

//--------------------------------------------------------------------------
- (NSString *) getHashCode
{
    return mHashCode;
}

@end

File: WeChatService/EX02-CSTSv3-ClientAPI/src/jedi/ex01/CSTS3/proxy/comm/CSTSTradeWaitObject.java
package jedi.ex01.CSTS3.proxy.comm;

import jedi.ex01.CSTS3.comm.ipop.IPFather;
import jedi.ex01.CSTS3.comm.ipop.OPFather;

public class CSTSTradeWaitObject {
	private final static long WAITGAP = 3 * 60 * 1000;
	private String _hashCode;
	private IPFather ip;
	private OPFather op;

	public CSTSTradeWaitObject(IPFather ip) {
		this.ip = ip;
		this._hashCode = ip.getOperateId();
	}

	public synchronized OPFather waitTrade(long waitTime) {
		if (op != null) {
			return op;
		}
		try {
			this.wait(waitTime);
		} catch (Exception e) {
			e.printStackTrace();
		}
		if (op == null) {
			OPFather op = new OPFather(ip);
			op.setSucceed(false);
			op
					.setErrCode(IPOPErrCodeTable.ERR_Timeout);
			op.setErrMessage("CSTS trade timedout!");
			return op;
		}
		return op;
	}

	public synchronized OPFather waitTrade() {
		return waitTrade(WAITGAP);
	}

	public synchronized void tradeReturn(OPFather op) {
		this.op = op;
		this.notifyAll();
	}

	public synchronized void setErr() {
		if (op != null) {
			this.notifyAll();
		} else {
			op = new OPFather(ip);
			op.setSucceed(false);
			op.setErrCode(IPOPErrCodeTable.ERR_NetErr);
			op.setErrMessage("CSTS net err!");
			tradeReturn(op);
		}
	}

	public String get_hashCode() {
		return _hashCode;
	}

}
